The plugin alias should not be ignored

AD7six opened the pull request:
I realize this might be contested (since it was intentional functionality), but allowing fuzzy-alias matching is problematic, and IMO propogating developer error - which the framework shouldn't be doing.

The scenario which brought me here was this:

![gc](https://cloud.githubusercontent.com/assets/33387/6059085/f73b2d70-ad2d-11e4-93b6-ed7a5013f17f.png)

The error actually came from debug kit:

```
2015-02-05 10:22:07 Error: [BadMethodCallException] Unknown method "gc"
Request URL: /
Stack Trace:
#0 vendor/cakephp/debug_kit/src/Routing/Filter/DebugBarFilter.php(187): Cake\ORM\Table->__call('gc', Array)
#1 vendor/cakephp/debug_kit/src/Routing/Filter/DebugBarFilter.php(187): Api\Model\Table\RequestsTable->gc()
#2 vendor/cakephp/cakephp/src/Event/EventManager.php(380): DebugKit\Routing\Filter\DebugBarFilter->afterDispatch(Object(Cake\Event\Event), Object(Cake\Network\Request), Object(Cake\Network\Response))
#3 vendor/cakephp/cakephp/src/Event/EventManager.php(346): Cake\Event\EventManager->_callListener(Array, Object(Cake\Event\Event))
#4 vendor/cakephp/cakephp/src/Event/EventManagerTrait.php(78): Cake\Event\EventManager->dispatch(Object(Cake\Event\Event))
#5 vendor/cakephp/cakephp/src/Routing/Dispatcher.php(92): Cake\Routing\Dispatcher->dispatchEvent('Dispatcher.afte...', Array)
#6 webroot/index.php(37): Cake\Routing\Dispatcher->dispatch(Object(Cake\Network\Request), Object(Cake\Network\Response))
#7 {main}
```

Because [these lines of code](https://github.com/cakephp/debug_kit/blob/3.0/src/Routing/Filter/DebugBarFilter.php#L186-L187):

```
$requests = TableRegistry::get('DebugKit.Requests');
$requests->gc();
```

Did _not_ return the table class from debug kit, but instead from my own Api plugin. I would expect this to be an ongoing problem since namespaces aught to eliminate the need for loose-alias matching.

So, should we use table aliases plugin-prefixed within the table registry?

AD7six committed:
The plugin alias should not be ignored

This allows unexpected scenarios such as:

    $wrongPluginModel = TableRegistry::get('This.Foo');

lorenzo commented:
I'm not against this, but I think it needs a fair amount of integration testing to make sure it won't be a hell to use in real life. Like the automatic table instantiation in controllers and shells.

AD7six commented:
I'll flesh-out the tests and see if I hit any problems.

AD7six committed:
Make load model more precise

If it's a plugin model, use the model alias to check for an existing
property

I.e. prevent this:

    $this->loadModel('Foo.Comments');
    $this->loadModel('Bar.Comments');  // Will overwrite $this->Comments
    $this->loadModel('Comments');  // Will NOT overwrite $this->Comments

If a user were to use referenes to multiple models with the same alias,
the first one should always be returned.

AD7six committed:
Add a test for load model being plugin-prefixed

lorenzo commented on tests/TestCase/ORM/TableRegistryTest.php:
```diff
@@ -376,14 +368,23 @@ public function testRemove()
-        $this->assertTrue(TableRegistry::exists('Comments'));
-        $this->assertNotSame($pluginTable, $appTable);
+        $pluginTable = TableRegistry::get('TestPlugin.Comments');
+        $cachedTable = TableRegistry::get('Comments');
```
Should you assert something here?

AD7six replied:
I think I'll move the test/asserts around to be more descriptive - rather than the current "make it work" state.

AD7six committed:
Add a test for overlapping aliases

AD7six committed:
Add convirmatory exists test

AD7six committed:
Add a few more asserts

AD7six committed:
Make testRemove meaningful

AD7six committed:
add a more comprehensive plugin-delete test

AD7six committed:
Whitespace fix

markstory commented on src/ORM/TableRegistry.php:
```diff
@@ -154,9 +153,9 @@ public static function config($alias = null, $options = null)
+    public static function get($alias, array $options = [])
     {
-        list(, $alias) = pluginSplit($name);
+        list(, $classAlias) = pluginSplit($alias);
```
Why do you split here, but not in exists()?

AD7six replied:
Only to get the class alias, used later in the method.

jadb commented:
Nice!

markstory commented:
I wonder if/how much chaos this will cause for plugins that already exist. I agree with the change in principle. The original issue for this change was #3425 and #3430 was the code changes. It looks like these changes shouldn't cause that issue to regress as the className option is being handled properly.

ADmad commented:
I like this :+1:

AD7six committed:
Flatten the code

Remove used-once intermediary variables, and move the pluginSplit call
to where it's used

AD7six committed:
Add a test demonstrating current behavior

The association name is used to store the object in the table registry

AD7six commented:
I think there's definitely going to be some chaos.

@markstory I'm a bit fuzzy on #3425; should we be deriving the table-registry key from the className? The tests added in #3430 only demonstrate that the table registry works as expected in this regard (which of course I've now changed), rather than the association class. 

[This test](https://github.com/cakephp/cakephp/commit/f575683ebf01b1c09ac96df69b139e6a52587bbe) suggests to me that the association class relies on the previous plugin-swallowing behavior and the assumption that the association name will be the same as the table alias (more or less).

AD7six committed:
oops

AD7six committed:
Demonstrate that there isn't alias-confusion

When dealing with overlapping table aliases

AD7six committed:
Better example test

AD7six committed:
Remove debug code

AD7six committed:
Remove nesting

The only way the code can arrive there, is if $table is null

AD7six committed:
Use the class name as a table alias

AD7six committed:
Flatten/simplify

markstory commented on tests/TestCase/ORM/AssociationTest.php:
```diff
@@ -170,6 +171,43 @@ public function testTarget()
+                '_options', 'attachTo', '_joinCondition', 'cascadeDelete', 'isOwningSide',
+                'saveAssociated', 'eagerLoader', 'type'
+            ],
+            ['ThisAssociationName', $config]
```
Why are you using a mock here?

AD7six replied:
[Association is an abstract class](https://github.com/cakephp/cakephp/blob/3.0/src/ORM/Association.php#L31) - so it can't be instanciated directly, rather than create an empty double, I used a mock.

The mocked methods are [copied from the startup method](https://github.com/cakephp/cakephp/blob/3.0-table-registry-alias/tests/TestCase/ORM/AssociationTest.php#L65-L72) - which I should get rid of though.

AD7six committed:
It's only necessary to override abstract methods

So, don't just copy and paste the setup method call

AD7six committed:
Add RepositoryAlias

This is an attempt to resolve the problem of entity source being
ambiguous.

This permits (though does not ensure) that an entity has the means to
say which table object it came from. alias and repository alias are not
the same thing, repository alias is plugin-prefixed, whereas alias is
not.

AD7six commented:
I've added a potential solution to `EntityTrait::source` being inaccurate - I think that's the only sticking point for this PR.

AD7six committed:
Don't inflect the alias

If it's wrong, it's not likely to be more-right when it's inflected

markstory commented on src/ORM/Table.php:
```diff
@@ -349,6 +359,23 @@ public function alias($alias = null)
+     * @param string|null $repositoryAlias the key used to access this object
+     * @return string
+     */
+    public function repositoryAlias($repositoryAlias = null)
```
registryAlias()?

markstory commented:
This is looking good. While I have some concerns that the stricter aliasing will cause issues, I think they will be good issues that help people in the long run. I know a few people have had complaints about the alias singletons before.

AD7six committed:
Rename repositoryAlias to registryAlias

it refers to the key used in the table _registry_ to access the relevant
table object, rather than the alias of the table itself

AD7six committed:
The table registry needs to pass the same option

Otherwise the table will ignore it.

lorenzo commented:
I like this. Good Job @AD7six !

Lennaert commented:
I think this pull-request introduced some unexpected behaviour to the assocations.

For example if I follow the 3rd example in the book http://book.cakephp.org/3.0/en/orm/associations.html ___(The same table can be used multiple times to define different types of associations.)___ and try to have an alias with ```$this->hasMany('UnapprovedComments', ['className' => 'Comments']);```

Then $this->Articles->UnapprovedComments->target() shows: 
```'alias' => 'Comments',```

and $this->Articles->find()->contain(['UnapprovedComments'])->all() shows:
 ```'unnaproved_comments' => null, 'Comments' => ['id' => '1','article_id' => '1']```

patrickconroy commented:
This also introduced a bug with plugin i18n, which I'm not sure how the tests didn't catch (it's totally broken for me now). `find('translations')` now produces:
``Fatal error: [PDOException] SQLSTATE[42S22]: Column not found: 1054 Unknown column 'Plugin_I18n.model' in 'where clause'``


In `TranslateBehavior::setupFieldAssociations` I had to change 
```
$this->_table->hasMany($targetAlias, [
    'className' => $table,
    'foreignKey' => 'foreign_key',
    'strategy' => $strategy,
    'conditions' => ["$targetAlias.model" => $fieldConditions['model']],
    'propertyName' => '_i18n',
    'dependent' => true
]);
```

to
```
$this->_table->hasMany($targetAlias, [
    'className' => $table,
    'foreignKey' => 'foreign_key',
    'strategy' => $strategy,
    'conditions' => ["model" => $fieldConditions['model']],
    'propertyName' => '_i18n',
    'dependent' => true
]);
```

And things *seem* to work. Should I make a PR/test? Or is there a bug with the way that association is working? I'd expect to use conditions keyd off the association name, right?

markstory commented:
@Lennaert I would say that is unexpected :cry:

AD7six commented:
Well, that didn't take long.

Will have a look why, and why the existing tests didn't pick that up.

patrickconroy commented:
Also, if a form accepts an Entity derived from a plugin in the create method, the core will now wrongly attempt to get the Table reference by the alias.

```
$thing = TableRegistry::get('Plugin.Stuff')->newEntity();
$this->Form->create($thing);
```

Will eventually do this, 
``
#1 ..app/vendor/cakephp/cakephp/src/View/Form/EntityContext.php(136): Cake\ORM\TableRegistry::get('Stuff')
``

Which of course won't work, it needs to be 
``
Cake\ORM\TableRegistry::get('Plugin.Stuff')
``

patrickconroy commented:
After debugging, it seems that forms will work if you set the context / table:

``
echo $this->Form->create($video, ['context' => ['table' => 'Plugin.Videos']]);
``

Is that the new normal? The `Entity->source` is just `Videos`, but that doesn't work. Should I attempt a fix to make the `Entity->source` return the string that's expected from the TR?

markstory commented:
@patrickconroy I don't think that should be required. I believe @AD7six is looking into ways to fix that. If we can't fix it, I'd prefer to revert the changes that caused all the problems.

AD7six commented:
I'd rather we didn't revert, this pr isn't in RC2, and while it sure is a PITA we/I aught to be able to sort them out fast enough. All these errors are implicit "source == table registry key" assumptions I missed in this PR, and we won't find them without exposing ourselves to a brief, difficult, period of chaos.

@patrickconroy best action for any such scenario is to PR a failing test case (without making any assumptions about how the internals work), though I would expect https://github.com/cakephp/cakephp/pull/5884 fixes that particular problem.

markstory commented:
Fair enough @AD7six .

AD7six commented:
I think everything mentioned here has been fixed now - should anyone land here with a related problem please create a separate issue for it so it doesn't get lost in the noise.
